mautic/mautic
# Tracking by lead device

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | Y
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Tracking logic refactored. Since now, contact was tracked in 2 steps
1. get tracking cookie - get or create new cookie with name mautic_session_id (mtc_sid on 3rd party sites) and assign it value of random hash
2. track current lead - create new cookie with name of mautic_session_id value and assign it lead id (mtc_id cookie on 3rd party sites).
Contact was than tracked by Lead ID.

New tracking create new cookie mautic_device_id (same name for 3rd party sites) with unique random string and this value is saved to LeadDevice entity associated with Lead. If cookie doesn't exist or no entity with tracking string is found, Lead is not identified.

Only first device that will access site with deprecated Mautic tracking cookies for Lead will be identified as that Lead. Any other will be created as new LeadDevice with new Lead until identified and merged.

[//]: # ( As applicable: )

#### Steps to test this PR:
1. Checkout on staging
2. Open two different browsers (or on two different devices) Mautic tracked page
3. Make both of them tracked as same Lead (by identification)
4. Checkout on PR branch
5. Access site from one of devices/browsers
6. New cookie called mautic_device_id should appear
7. New LeadDevice entity should appear with tracking id value same as in cookie mautic_device_id
8. New device should be associated with previous Lead
8. Access site from other device/browser
9. New cookie called mautic_device_id should appear
10. New LeadDevice entity should appear with tracking id value same as in cookie mautic_device_id
11. New device should be associated with newly created Lead
12. If other device is identified as previous Lead, it should be merged to one (and other device should get new Lead id)
13. Remove all cookies
14. Make action that tracks Lead
15. Cookie should mautic_device_id should be created with new Lead and both Entities should be in database (Lead and LeadDevice)

It should work in hit page, download asset, form submission, sites with included mtc.js and basically everywhere where is LeadModel::getCurrentLead used.

#### List backwards compatibility breaks:
**LeadModel::getTrackingCookie (removed)**
_Accessible from ContactTrackingService::getTrackingIdentifier with similar functionality, however it's deprecated so better is to reimplement with DeviceTrackingService. LeadDevice Entity now has attribute trackingId accessible by LeadDevice::getTrackingId()_

**LeadModel::trackCurrentLead (removed)**
_Similar as above - tracking logic was refactored_

**LeadModel::getCurrentLead, getSystemCurrentLead, setCurrentLead and setSystemCurrentLead (deprecated)**
Use ContactTracker::getTrackedContact, getSystemContact, setTrackedContact, and setSystemContact instead.
